File: 1727-largest-submatrix-with-rearrangements/1727-largest-submatrix-with-rearrangements.cpp
class Solution {
public:
    int largestSubmatrix(vector<vector<int>>& matrix) {
        int res = 0;
        int m = matrix.size();
        int n = matrix[0].size();
        vector<pair<int, int>> prevHeights;
        for(int i = 0; i < m; i++)
        {
            vector<pair<int, int>> heights;
            vector<bool> seen(n, false);
            for(int j = 0; j < prevHeights.size(); j++)
            {
                if(matrix[i][prevHeights[j].second] == 1)
                {
                    heights.push_back({prevHeights[j].first + 1, prevHeights[j].second});
                    seen[prevHeights[j].second] = true;
                }
            }
            for(int j = 0; j < n; j++)
            {
                if(!seen[j] && matrix[i][j] == 1)
                {
                    heights.push_back({1, j});
                }
            }
            int sumHeight = 0;
            for(int j = 0; j < heights.size(); j++)
            {
                res = max(res, heights[j].first * (j + 1));
            }
            prevHeights = heights;
        }
        return res;
    }
};

File: 1727-largest-submatrix-with-rearrangements/1727-largest-submatrix-with-rearrangements.py
class Solution:
    def largestSubmatrix(self, matrix: List[List[int]]) -> int:
        res = 0
        m = len(matrix)
        n = len(matrix[0])
        prevHeights = []
        for i in range(m):
            heights = []
            seen = [False for _ in range(n)]
            for j in range(len(prevHeights)):
                if matrix[i][prevHeights[j][1]] == 1:
                    heights.append([prevHeights[j][0] + 1, prevHeights[j][1]])
                    seen[prevHeights[j][1]] = True
            for j in range(n):
                if not seen[j] and matrix[i][j] == 1:
                    heights.append([1, j])
            for j in range(len(heights)):
                res = max(res, heights[j][0] * (j + 1))
            prevHeights = heights
        return res